Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch

```diff
@@ -0,0 +1,137 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Thonk <30448663+ExcessiveAmountsOfZombies@users.noreply.github.com>
+Date: Tue, 15 Dec 2020 09:07:01 -0600
+Subject: [PATCH] Allow individual plugins to load datapacks
+
+
+diff --git a/src/main/java/io/papermc/paper/datapack/PluginDataPack.java b/src/main/java/io/papermc/paper/datapack/PluginDataPack.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..50608567cd3f9d8822a85944669a2984314de1f8
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datapack/PluginDataPack.java
+@@ -0,0 +1,25 @@
++package io.papermc.paper.datapack;
++
++import net.minecraft.server.ResourcePackFile;
++
++import java.io.File;
++
++public class PluginDataPack extends ResourcePackFile {
++
++    private String name;
++    private boolean enabledByDefault;
++
++    public PluginDataPack(String name, File file, boolean enabledByDefault) {
++        super(file);
++        this.name = name;
++        this.enabledByDefault = enabledByDefault;
++    }
++
++    public String getName() {
++        return name;
++    }
++
++    public boolean isEnabledByDefault() {
++        return enabledByDefault;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/datapack/PluginSource.java b/src/main/java/io/papermc/paper/datapack/PluginSource.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..10a08118beda6b24533ae054f37acfb78d74bd7e
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datapack/PluginSource.java
+@@ -0,0 +1,46 @@
++package io.papermc.paper.datapack;
++
++import net.minecraft.server.EnumResourcePackType;
++import net.minecraft.server.PackSource;
++import net.minecraft.server.ResourcePackLoader;
++import net.minecraft.server.ResourcePackSource;
++import org.apache.commons.io.FilenameUtils;
++
++import java.io.File;
++import java.util.ArrayList;
++import java.util.List;
++import java.util.function.Consumer;
++
++public class PluginSource implements ResourcePackSource {
++
++    private File pluginFolder;
++
++    public PluginSource(File pluginFolder) {
++        this.pluginFolder = pluginFolder;
++    }
++
++    @Override
++    public void a(Consumer<ResourcePackLoader> consumer, ResourcePackLoader.a creator) {
++        List<PluginDataPack> packs = new ArrayList<>();
++
++        if (pluginFolder.exists()) {
++            for (File file : pluginFolder.listFiles((dir, name) -> name.endsWith(".jar"))) {
```

Comment:
Should every reload look for new files with the `.jar` name? I don't think new plugins' datapacks should be reloaded with /minecraft:reload if the server doesn't know about those plugins already. Maybe move this to constructor and store a list of PluginDataPacks.